src/pip/_internal/network/lazy_wheel.py

```diff
@@ -1,66 +1,69 @@
+from pip._vendor.requests.models import CONTENT_CHUNK_SIZE, HTTPError, Response
+from pip._vendor.requests.status_codes import codes
 
+from pip._internal.exceptions import InvalidWheel, UnsupportedWheel
 from pip._internal.metadata import BaseDistribution, MemoryWheel, get_wheel_distribution
-from pip._internal.network.session import PipSession
-from pip._internal.network.utils import HEADERS, raise_for_status, response_chunks
+from pip._internal.network.session import PipSession as Session
+from pip._internal.network.utils import HEADERS
+from pip._internal.utils.logging import indent_log
+
+logger = logging.getLogger(__name__)
 
 
 class HTTPRangeRequestUnsupported(Exception):
-    pass
+    """Raised when the remote server appears unable to support byte ranges."""
 
 
-def dist_from_wheel_url(name: str, url: str, session: PipSession) -> BaseDistribution:
+def dist_from_wheel_url(name: str, url: str, session: Session) -> BaseDistribution:
     """Return a distribution object from the given wheel URL.
 
     This uses HTTP range requests to only fetch the portion of the wheel
     containing metadata, just enough for the object to be constructed.
-    If such requests are not supported, HTTPRangeRequestUnsupported
-    is raised.
-    """
-    with LazyZipOverHTTP(url, session) as zf:
-        # For read-only ZIP files, ZipFile only needs methods read,
-        # seek, seekable and tell, not the whole IO protocol.
-        wheel = MemoryWheel(zf.name, zf)  # type: ignore
-        # After context manager exit, wheel.name
-        # is an invalid file by intention.
-        return get_wheel_distribution(wheel, canonicalize_name(name))
 
-
-class LazyZipOverHTTP:
-    """File-like object mapped to a ZIP file over HTTP.
-
-    This uses HTTP range requests to lazily fetch the file's content,
-    which is supposed to be fed to ZipFile.  If such requests are not
-    supported by the server, raise HTTPRangeRequestUnsupported
-    during initialization.
+    :raises HTTPRangeRequestUnsupported: if range requests are unsupported for ``url``.
+    :raises InvalidWheel: if the zip file contents could not be parsed.
+    """
+    try:
+        # After context manager exit, wheel.name will point to a deleted file path.
+        # Add `delete_backing_file=False` to disable this for debugging.
+        with LazyWheelOverHTTP(url, session) as lazy_file:
+            lazy_file.prefetch_contiguous_dist_info(name)
+
+            wheel = MemoryWheel(lazy_file.name, lazy_file)
+            return get_wheel_distribution(wheel, canonicalize_name(name))
+    except (BadZipFile, UnsupportedWheel):
+        # We assume that these errors have occurred because the wheel contents themself
+        # are invalid, not because we've messed up our bookkeeping and produced an
+        # invalid file that pip would otherwise process normally.
+        raise InvalidWheel(url, name)
+
+
+class ReadOnlyIOWrapper(BinaryIO):
```

Comment:
We probably should have unit tests covering this and/or LazyWheelOverHTTP.

Reply:
Oh yes, I was going to do that earlier and forgot! Thanks!

Reply:
Just spent some time splitting out the merging logic into a separate `MergeIntervals` class to make it easier to unit test.